restapi/src/main/java/io/stargate/web/docsapi/service/query/search/resolver/impl/CnfResolver.java

```diff
@@ -0,0 +1,247 @@
+/*
+ * Copyright The Stargate Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package io.stargate.web.docsapi.service.query.search.resolver.impl;
+
+import com.bpodgursky.jbool_expressions.And;
+import com.bpodgursky.jbool_expressions.Expression;
+import com.bpodgursky.jbool_expressions.Literal;
+import com.bpodgursky.jbool_expressions.options.ExprOptions;
+import com.bpodgursky.jbool_expressions.rules.Rule;
+import com.bpodgursky.jbool_expressions.rules.RuleList;
+import com.bpodgursky.jbool_expressions.rules.RulesHelper;
+import com.google.common.collect.Multimap;
+import com.google.common.collect.Multimaps;
+import io.reactivex.rxjava3.core.Flowable;
+import io.stargate.web.docsapi.dao.DocumentDB;
+import io.stargate.web.docsapi.dao.Paginator;
+import io.stargate.web.docsapi.service.DocsApiConfiguration;
+import io.stargate.web.docsapi.service.ExecutionContext;
+import io.stargate.web.docsapi.service.RawDocument;
+import io.stargate.web.docsapi.service.query.FilterExpression;
+import io.stargate.web.docsapi.service.query.FilterPath;
+import io.stargate.web.docsapi.service.query.rules.TrueFilterExpressions;
+import io.stargate.web.docsapi.service.query.search.resolver.BaseResolver;
+import io.stargate.web.docsapi.service.query.search.resolver.DocumentsResolver;
+import io.stargate.web.docsapi.service.query.search.resolver.filter.CandidatesFilter;
+import io.stargate.web.docsapi.service.query.search.resolver.filter.impl.InMemoryCandidatesFilter;
+import io.stargate.web.docsapi.service.query.search.resolver.filter.impl.PersistenceCandidatesFilter;
+import io.stargate.web.docsapi.service.query.search.weigth.ExpressionWeightResolver;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Optional;
+import java.util.function.Function;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+
+/** Resolver that accepts the {@link And} expression in the CNF form. */
+public class CnfResolver implements DocumentsResolver {
+
+  private final DocumentsResolver delegate;
+
+  public CnfResolver(
+      Expression<FilterExpression> expression,
+      ExpressionWeightResolver<FilterExpression> weightResolver,
+      ExecutionContext context,
+      DocumentsResolver parent) {
+    if (!And.EXPR_TYPE.equals(expression.getExprType())) {
+      throw new IllegalArgumentException(
+          "CNF resolver expect the AND expression in the conjunctive normal form.");
```

Comment:
yea, but I need to check how to prove that it's in the CNF form.. suggestions?

Reply:
Sorry, no particular suggestion at this point. If you're ok with refactoring all `resolve` logic into a separate call chain (i.e. remove `delegate` and deal with all expression types in static methods or even in a separate utility class) this check may no longer be necessary... rather it may naturally become a branch in an `if` or `switch` and not require an exception :thinking:

Reply:
Delegate removed :+1: